core/trino-main/src/test/java/io/trino/sql/planner/assertions/JoinMatcher.java

```diff
@@ -131,6 +135,11 @@ public MatchResult detailMatches(PlanNode node, StatsProvider stats, Session ses
             return NO_MATCH;
         }
 
+        if (taskRetriesEnabled && !joinNode.getDynamicFilters().isEmpty()) {
```

Comment:
It seems like the `JoinMatcher` / `SemiJoinMatcher` is a generic class and can be used at the arbitrary step of planning process. This assumption looks weird here. Shouldn't it be covered by the `DynamicFilterPattern` that is set accordingly to the expectations?